corgi/collectors/pulp.py

```diff
@@ -0,0 +1,108 @@
+        "limit": PAGE_SIZE,
+        "skip": 0,
+    }
+}
+
+MODULE_CRITERIA = {
+    "criteria": {
+        "fields": {"unit": ["name", "stream", "artifacts", "version", "context", "arch"]},
+        "type_ids": ["modulemd"],
+    }
+}
+
+
+class Pulp:
+    def __init__(self):
+        self.session = requests.Session()
+
+    def get_active_repositories(self) -> int:
+        """Fetch all Pulp repos which had something shipped in them, and their content sets."""
+        criteria = {
+            "criteria": {
+                "fields": ["id", "notes"],
+                "filters": {"last_unit_added": {"$ne": None}},
+            }
+        }
+        url = f"{settings.PULP_URL}/api/v2/repositories/search/"
+        auth = (settings.PULP_USERNAME, settings.PULP_PASSWORD)
+        response = self.session.post(url, json=criteria, auth=auth)
+        response.raise_for_status()
+        no_created = 0
+        for repo in response.json():
+            content_set = repo["notes"].get("content_set", "")
+            _, created = CollectorRPMRepository.objects.update_or_create(
+                name=repo["id"], defaults={"content_set": content_set}
+            )
+            if created:
+                no_created += 1
+                logger.info("Created CollectorRPMRepository with id %s", repo["id"])
+        return no_created
+
+    def get_module_data(self, repo: str) -> Generator[str, None, None]:
+        module_data = self._get_module_data(repo)
+        yield from Brew().persist_modules(module_data)
+
+    def _get_module_data(self, repo) -> dict[str, list[str]]:
+        module_data = self._get_unit_data(repo, MODULE_CRITERIA)
+        rpms_by_module = {}
+        for entry in module_data:
+            entry_meta = entry.get("metadata")
+            name = entry_meta.get("name")
+            # eg. build_id 1500134
+            stream = entry_meta.get("stream").replace("-", "_")
+            version = entry_meta.get("version")
+            context = entry_meta.get("context")
+            module_key = f"{name}-{stream}-{version}.{context}"
+            rpms_by_module[module_key] = entry_meta.get("artifacts")
+        return rpms_by_module
+
+    def get_rpm_data(self, repo: str) -> Generator[str, None, None]:
+        rpms_by_srpm = self._get_rpm_data(repo)
+        yield from Brew().lookup_build_ids(rpms_by_srpm)
+
+    def _get_rpm_data(self, repo) -> defaultdict:
+        rpms_by_srpm = defaultdict(list)
+        page = 0
+        while True:
+            rpm_criteria = copy.deepcopy(RPM_CRITERIA)
+            rpm_criteria["criteria"]["skip"] = PAGE_SIZE * page
+            page += 1
+            rpm_data = self._get_unit_data(repo, rpm_criteria)
+            for entry in rpm_data:
+                entry_meta = entry.get("metadata")
+                filename = entry_meta.get("filename")
+                source_rpm = entry_meta.get("sourcerpm")
+                rpms_by_srpm[source_rpm.removesuffix(".rpm")].append(filename)
```

Comment:
Are we using .get() here because these keys might not be present in the entry dict? If yes, we need to handle the case when entry_meta or source_rpm are None. With the current code, ```entry_meta.get("filename or sourcerpm")``` and ```source_rpm.removesuffix(",rpm")``` will fail when if either of those is None.

Or if not, and we expect those values to always be present:
```suggestion
            for entry in rpm_data:
                filename = entry["metadata"]["filename"]
                source_rpm = entry["metadata"]["sourcerpm"].removesuffix("rpm")
                rpms_by_srpm[source_rpm].append(filename)
```

Reply:
I don't expect them to be None, so using your suggestion.